File: 8Feb2024.py
from cmath import inf


class Solution:
    def numSquares(self, n: int) -> int:
        def f(v,dp) :
            if(v==0) :
                return 0
            if(dp[v]!=-1) :
                return dp[v]
            ans = inf
            for i in range(1,v+1) :
                if(i*i>v) :
                    break
                f1 = f(v-(i*i),dp)
                ans=min(ans,f1+1)
            dp[v]=ans
            return ans
        dp = [-1]*(n+1)
        return f(n,dp)

File: 8Feb2024.java
class Solution {
    public int numSquares(int n) {
        return f(n, new Integer[n + 5]);
    }

    public int f(int v, Integer[] dp) {
        if (v == 0)
            return 0;
        if (dp[v] != null)
            return dp[v];
        int ans = Integer.MAX_VALUE;
        for (int i = 1; i * i <= v; i++) {
            int f1 = f(v - (i * i), dp);
            ans = Math.min(ans, f1 + 1);
        }
        return dp[v] = ans;
    }
}